### ISSUE TITLE
yield-like API for simpler cross-system-call control-flow

### ISSUE THREAD

**Issue by devietti** [upenn-acg/detTrace#124]
Here's a proposal to make it simpler to write logic that extends across multiple system calls, e.g., when we need to inject a system call. The key abstraction is functionality like Python's "yield" statement, referred to here as "await".

Let's consider the case of `alarm` (`open`, which injects an `fstat`, is even more complicated but I'm less familiar with it). I want to be able to write code like this:

```
alarmPreHook(tracerObj, monitorState, processState) {

   int alarmInterval = getArg1();

   if (sigalrmHandler == IGNORE) {
      changeSystemCall( SYS_getpid ); // make this system call a NOP
      return; // don't wait for post-hook
   }
   if (sigalrmHandler == DEFAULT) {
      changeSystemCall( SYS_exit ); // the SIGALRM causes us to exit
      setArg1( -SIGALRM ); // act like we were killed by a signal
      awaitImmediatePreHook( SYS_exit ); // SYS_exit has no post-hook
      return;
   }
   if (sigalrmHandler == CUSTOM) {
      tgkill( thisTracee, SIGALRM );
      changeSystemCall( SYS_pause );
      awaitImmediateSignal( SIGALRM ); 
      awaitImmediatePostHook( SYS_pause ); 
      setReturnValue( 0 ); // pretend to be the end of original alarm() call
      return;
   }
}
```

This design takes a complicated thread of control, currently spread across the `alarm` and `pause` pre/post hooks, and condenses it into a single place. The `await*()` calls are like "yield" statements, and when execution resumes afterwards the arguments to the original `alarmPreHook` (`tracerObj, monitorState, processState`) are updated to reflect the current event being intercepted.

Semantically, I want to imagine a separate thread T to run this `alarmPreHook()`, and when T reaches an `await()` call it blocks and we transfer control back to the main event loop. When an event arrives, a dispatcher looks at all of the blocked threads to see which (if any) "match" the event.

These `await()` calls are limited in scope to the "current" tracee - "I" am waiting for an event from my current tracee, not from any other tracee. But perhaps a more advanced variant could wait for events from other tracees (e.g., wait for someone to try to send me a signal?).

`awaitImmediate()` is an assertion that the next event from my current tracee should be of the given kind. If I'm awaiting a `SYS_pause` pre-hook but I actually see a `SYS_read` pre-hook next, then that's an error -- I wasn't expecting a `SYS_read` to show up so I may need to do more work to handle this intermediate call.

Additionally, we could also have an `awaitEventual()` call, which says that it's ok for other syscalls to arrive from my tracee, but once I do receive the event I'm waiting for, fire up my thread again. Or we could have as rich a "pattern-matching" language as we want for expressing what we expect the event stream to be. I think there's an analogy to regular expressions for the structure of these event streams; not sure if that's worth developing more explicitly or not.

We'd have to think about how to treat signals. Should we always explicitly await their delivery, as done in the `alarm` example above? Maybe we need an `awaitImmediateExceptSignal( SYS_read )` that says: "the next thing from my tracee is either an arbitrary number of signals, or a SYS_read". In regex terms it would be `signal*read`. Certainly the arrival of a signal could generate other system calls (from within the signal handler)...

If there is no blocked thread that is awaiting the next event that arrives, then we spin up a new thread to handle it. This could occur when tracee A's thread is blocked and an event arrives for tracee B, or when tracee A is doing an `awaitEventually` and we want to handle other events from tracee A (such as the arrival of a signal, and its handler running) independently. For efficiency, we'd want to use a thread pool so we don't have to spin up new pthreads for each prehook.

### Advantages

One key advantage of this organization is that it is _reentrant_. If I'm waiting for a `SYS_alarm` from my tracee, and then another tracee does a `SYS_alarm`, I won't get confused. Each tracee will have their own thread following the logic above, and the events for each tracee are totally independent. Currently, such an interleaving of events would almost certainly corrupt state like `state::signalInjected` or `state::syscallInjected`.

Another advantage is that it would be easier to reason about and transfer state across system calls and hooks. The value of `sigalrmHandler` is available throughout several pre/post hooks, stored as a local variable on my thread's stack instead of in a global or per-process object. There's no worry about this state being corrupted by intermediate system calls, or forgetting to reset this state for the next run of the "state machine" (#119). Each pre-hook can accumulate whatever state it needs as it executes. Saving and restoring state around injected system calls would be much cleaner.

Code for each system call would be simpler. The common case of pre+post hook would be significantly condensed, and we only need to declare a preHook for system calls we want to intercept. Skipping the post-hook (or following a more complicated chain of logic beyond the post-hook) are all implicit in what the pre-hook awaits.

### Implementation

Most of the complexity lies in the "dispatcher" which gets ptrace events and matches them against blocked tracee threads. So we'd need to have a list of tracee threads and know what event(s) each is awaiting. If there are ever two blocked threads awaiting the same event that certainly seems very problematic, and should probably trigger a crash. There's always the last-resort mechanism to spawn a new thread, which has lower priority than an existing blocked thread. Perhaps we need richer priorities among the blocked threads as well in some cases? That would be kind of ugly.

For simplicity, there should only be one tracee thread ever actually running at a time. The dispatcher gets an event, decides which blocked thread T to hand it to (or to create a new thread), and then control is transferred to T until it awaits another event, which hands control back to the dispatcher. The dispatcher then blocks on the ptrace fd.

Using real threads makes this sequential execution non-trivial to achieve, and would require condition variables to wake/sleep threads. In reality, all we really need for each "thread" is a _continuation_ (essentially, a stack). Threads are likely more debugger-friendly, however. I don't think we really get any benefits from the parallelism of threads: we could have true parallelism among tracees _without_ true parallelism at the monitor. We could investigate true parallelism within the tracer, too, then these threads might actually run in parallel.

State (like `tracerObj, monitorState, processState`) needs to be updated before an `await()` call returns to a tracee thread. The dispatcher would do this. As long as the updates occur before the tracee thread is woken up, there are no consistency model issues.

We could naturally implement more efficient buffering of updates to tracee state (like registers). Currently we send them to the tracee after each update: if you write 5 tracee registers in a pre-hook, that's 5 `PTRACE_SETREGS` calls. Instead, we could do just one `PTRACE_SETREGS` at `await()` calls. This could be done in the current design as well.

It would also be nice to have a single place to declare new pre-hooks, instead of needing to plumb them through seccomp.cpp, execution.cpp, and dettraceSystemCall.[ch]pp. Ideally, you'd just declare a class and define ::alarmPreHook in dettraceSystemCalls.hpp and that'd be it. Maybe some cpp magic can help accomplish this, like a macro for the ctor that auto-registers with 1) seccomp and 2) a big hash table for matching syscalls to the proper pre-hook.

### Questions

One piece of DetTrace I don't understand is the tracee scheduler. How well would this design support the scheduler? Each tracee is potentially "spread" across multiple threads.
